src/app/Fake.Core.Target/Target.fs

```diff
@@ -444,11 +444,15 @@ module Target =
                     | Some e -> alignedError name time e.Message)
 
             aligned "Total:" total null
-            if not context.HasError then aligned "Status:" "Ok" null
+            if not context.HasError then 
+                aligned "Status:" "Ok" null
+                Trace.setBuildState TagStatus.Success
             else
                 alignedError "Status:" "Failure" null
+                Trace.setBuildState TagStatus.Failed
```

Comment:
@BlythMeister Is there any particular reason why we need that? I'm asking because I just realized (after trying to figure out for hours which commit is to blame) that this makes the VSTS build red. This is because we have some testing going on with negative/failing cases which will call this API and make the build red.

Reply:
The reason is so that when running on a CI (like teamcity) when the build fails the state is set to failure.
It's to try and prevent the build failure ending with "Process returned exit code 1"

Reply:
But shouldn't that `Process returned exit code 1` already be solved by using the error apis?

Reply:
possibly...happy to revert this bit of the change and try

Reply:
It does mean though that any listener with "TraceData.BuildState" is a bit pointless as nothing ever traces with that...

Reply:
No it is a public API which people or other helpers can use?
The other solution would need to be some "internal" & "undocumented" flag to disable this API for our bootstrapping process. I'd assume it is only a problem for us and not for our "normal" users.

The only problem that comes to mind would be nested fake scripts which are allowed to fail, so even those users probably need a way to disable that API anyway.

Reply:
I think adding a feature switch for this would be ideal.

How/Where should it be added (happy to do this :))

Reply:
No idea, I like having stuff explicit and without global mutable state. Usually that makes stuff hard ;)

Reply:
How about we make it so there is a new run or update existing runs to return the `TargetContext` from `runInternal`
That way, as a user you can make the decision to react to the overall run result...

Reply:
(facepalm)....a bit like `runAndGetContext`

I'll shut up :)

Reply:
```
match context.PreviousTargets.Length, context.HasError with
        | 0, _ -> Trace.setBuildState TagStatus.Warning
        | _, true -> Trace.setBuildState TagStatus.Failed
        | _, _ -> Trace.setBuildState TagStatus.Success
```
in my script after calling `runAndGetContext` means it does what i want :)

Reply:
Yes I should have been a bit more conservative with these changes... Found another edge case where the build went red in teamcity (good that we have a teamcity build now) so I need a bit more time to release this properly.

Regarding your suggestion. Yes indeed we could make that an opt-in api:

```fsharp
let context = Target.tryRunOrDefaultAndGetContext "Default"
Target.reportBuildStateFromContext context
```

(Or something along those lines)

Can you open a PR or an issue for this (so it is not forgotten in this PR)

Reply:
I'll take a look next week and ensure that all run methods also have a WithContext
and add a function to report status based on context which you could (and I will) pipe the run into.
Watch this space 😂

Reply:
I'll take a look next week and ensure that all run methods also have a WithContext
and add a function to report status based on context which you could (and I will) pipe the run into.
Watch this space 😂